### ISSUE TITLE
Wrong units for DNB-based composites

### ISSUE THREAD

**Issue by gerritholl** [pytroll/satpy#1985]
**Describe the bug**

There are several DNB-based single-channel composites built into Satpy: adaptive_dnb, histogram_dnb, dynamic_dnb, and hncc_dnb.  They all do more or less complex operations on the calibrated radiances.  I believe those are all intended for visualisation purposes and not for further processing.

Those more or less complex operations have impacts on the unit of the data.  The resulting value can no longer be directly interpreted as a radiance.  Perhaps the correct units include a sqrt, a log, or no unit can be assigned at all.

However, in reality, the `hncc_dnb` composite claims to present radiances in W m¯² sr¯¹, whereas the histogram_dnb, adaptive_dnb, and dynamic_dnb all claim to present radiances in units 1.  Radiance cannot have unit 1, and I think the `hncc_dnb` composite takes the square root of the data, so in all cases the units are wrong.

**To Reproduce**

```python
from satpy import Scene
from glob import glob
#from satpy.utils import debug_on; debug_on()
filenames = glob("/media/nas/x21308/scratch/VIIRS/npp_20220119_0132_53000/*DNB*.h5")
sc = Scene(filenames=filenames, reader=["viirs_sdr"])
names = ["DNB", "hncc_dnb", "histogram_dnb", "adaptive_dnb", "dynamic_dnb"]
sc.load(names)
for name in names:
    print(name, sc[name].attrs["calibration"], sc[name].attrs["units"])
```

**Expected behavior**

I expect that DNB has units of W m¯² sr¯¹, but the others should all have different units or no units at all.

**Actual results**

The units are expected for DNB but unexpected for all others.

```
DNB radiance W m-2 sr-1
hncc_dnb radiance W m-2 sr-1
histogram_dnb radiance 1
adaptive_dnb radiance 1
dynamic_dnb radiance 1
```

**Environment Info:**
 - OS: openSUSE 15.3
 - Satpy Version: v0.33.1-39-gcccbe06f

**Additional context**

The unit `1` causes problems in pyninjotiff, which tries and fails to convert it to `%` or `N/A`, depending on what I configure.  As a consequence, pyninjotiff cannot write the DNB composites.

The wrong units also appear to affect composites.  When I load `true_color`, the metadata also state `calibration: 'radiance'` and `units: '%'`.  I'm not sure why in this case pyninjotiff has no problems writing.

**Comment by djhoese** [comment]
The replacement of those units with "1" was meant to signify that these are no longer radiance. I believe the standard name is also overwritten to "normalized_radiance". What else is saying they are a radiance? The calibration? Similar to resolution remaining the same even after resampling, maybe calibration can remain too?

**Comment by gerritholl** [comment]
Yes, the `standard_name` has changed.  I think the most correct would be to remove the `units` attribute altogether or set it to `None`, if it cannot be meaningfully defined.  The value `1` is problematic because it's a valid unit for reflectance.  I think that in one of our old VIIRS productions, an old version of pyninjotiff uses `pint` to convert the normalised or equalized radiance from units `1` to units `%`, because our (wrong) configuration indicated `%` as the output unit for DNB images.  I replaced a not-working unit conversion by a NotImplementedError in #1615.

**Comment by djhoese** [comment]
Back when I implemented the ninjo writer (backend) in Polar2Grid for Katja at DWD, the units were 1 as they are now for these normalized DNB products. They may have shown `%` in the GUI and we just let that slide, but I wasn't aware of it. I have generally used `1` as the "this doesn't have any other meaningful unit". I'm not saying that is correct, but I don't think have no units will work either. For example, I think AWIPS (the US forecasting GUI) requires units to be specified so that's another place that `1` is used. Since `1` is a valid unit for all ratios/fractions, it should not be considered a reflectance any time it is seen.

**Comment by gerritholl** [comment]
When I produce RGBs for NinJo, I'm setting units attribute to `"N/A"`.  In the Satpy ninjotiff writer, there is a guard in which `convert_units` is only called when the `units` attribute is set.  I tried to do the same for the DNB composites, but for a reason I don't completely understand yet, this fails with `unit conversion only implemented between K and C, not between 1 and N/A`.  I'm not sure why I don't have this problem for regular RGBs, which in some cases also still have a `units` attribute.

I agree that `1` should not be automatically considered a reflectance; but the fact that it's valid for 'reflectance' is probably why our old production is working at all, by accident.  Our old production is also buggy.  My attempt to fix it I ran into an error message from the ninjotiff writer that "unit conversion only implemented between K and C, not between 1 and %", which is something I introduced myself in #1615 because I saw that the old implementation was wrong.

**Comment by mraspaud** [comment]
CF proposes either "1" or "" for unitless quantities

**Comment by gerritholl** [comment]
Yes, but is `hncc_dnb` a quantity?

**Comment by djhoese** [comment]
I think we'd have to go to the referenced paper on HNCC. The histogram and adaptive are two types of histogram equalization. I would say a "normalization", especially a non-linear one, is no longer the original quantity in any way.

**Comment by gerritholl** [comment]
From [the HNCC paper](https://www.tandfonline.com/doi/full/10.1080/01431161.2017.1338838), although the gain factor is a complicated calculation, it is itself unitless, so does not change the units of the radiance.  So I guess that it's still the physically correct unit for the `ncc_radiance`, even if this is not a quantity whose values can be meaningfully interpreted.

**Comment by djhoese** [comment]
So where are we with this? I think based on the comments and hoping for some level of CF compliance the units `"1"` are valid units for any unitless quantity. I'm not sure if a normalized dataset is considered unitless. I think we've now agreed that the units for `hncc_dnb` are valid.

So if the main issue is that NinJo doesn't like it, is there a work around we can implement in the writer (as it sounds like you've tried at least in one of the writers)?

**Comment by gerritholl** [comment]
The unit "1" is correct for a unitless quantity, but `adaptive_dnb` and `histogram_dnb` are not quantities.

The CF conventions do not define what a quantity is, but ISO 80000-1 does:

> **quantity**  
> property of a phenomenon, body, or substance, where the property has a magnitude that can be expressed by means of a number and a reference

The `adaptive_dnb` and `histogram_dnb` composites apply different calculations in different parts of the image.  The resulting image is useful for visual interpretation, but like RGBs or enhanced images, the values no longer represent a property that can be expressed by a number and a reference.

ISO 80000 states that quantities must always have units associated with them.  Unfortunately, the CF conventions state that "Units are not required for dimensionless quantities".  This means not only that the CF conventions allow storing data in a way that violates ISO 80000, but also that we can no longer disambiguate between variables representing dimensionless quantities and variables that do not represent quantities at all.

ISO 80000 covers quantities and units.  It does not cover non-quantities.  In the CF conventions, I cannot find anything about non-quantities either.  Neither standard provides us guidance on what to do for non-quantities such as adaptive DNB, histogram DNB, or RGBs.  I don't know if there is another standard that can help us here.  I don't know if CF compliance is possible for non-quantities.

I would argue that the presence of the units attribute gives the impression that the associated variable represents a quantity (the same for #2066).  This is wrong for adaptive DNB and histogram DNB (and for RGBs).  Therefore, I think those should not have units at all.

**Comment by djhoese** [comment]
Wow, thanks for all that information. That's good to know. I wouldn't be surprised if the CF conventions were written without "image data" in mind. I think I agree with you that the `units` can be removed. I'll have to double check how I handle missing units in the AWIPS tiled writer. We may also have to make sure we document in the `readers.rst` document that `units` is optional and may not make sense for all data.

Should we add this and the RGB units to the next monthly meeting or do you need a resolution on this problem sooner than that?
